Moving to use gitignore library go-gitignore

mohammedzee1000 opened the pull request:
Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

**What type of PR is this?**
/kind bug

**What does this PR do / why we need it**:
See topic

**Which issue(s) this PR fixes**:

Fixes #4389

**PR acceptance criteria**:

- [x] Unit test 

- [ ] Integration test 

- [ ] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
```bash
$ git clone https://github.com/scottkurz/multi-module-devfile.git && cd multi-module-devfile
$ mvn install
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] guide-maven-multimodules-jar                                       [jar]
[INFO] guide-maven-multimodules-war                                       [war]
[INFO] guide-maven-multimodules-ear                                       [ear]
[INFO] guide-maven-multimodules                                           [pom]
[INFO] 
...
[INFO] Reactor Summary for guide-maven-multimodules 1.0-SNAPSHOT:
[INFO] 
[INFO] guide-maven-multimodules-jar ....................... SUCCESS [  1.139 s]
[INFO] guide-maven-multimodules-war ....................... SUCCESS [  0.422 s]
[INFO] guide-maven-multimodules-ear ....................... SUCCESS [  0.631 s]
[INFO] guide-maven-multimodules ........................... SUCCESS [  0.004 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.281 s
[INFO] Finished at: 2022-03-10T12:08:33+05:30
[INFO] ------------------------------------------------------------------------
$ $ cat .gitignore 
target/
.settings/
.project
.classpath
.factorypath
MANIFEST.MF
README.html
.DS_Store
bin/
build/
$ ls ear/target jar/target
ear/target:
application.xml               guide-maven-multimodules-ear.ear  test-classes
generated-test-sources        maven-archiver
guide-maven-multimodules-ear  maven-status

jar/target:
classes            guide-maven-multimodules-jar-1.0-SNAPSHOT.jar  maven-status
generated-sources  maven-archiver
$ odo push -v 7
0310 12:22:47.356831   17253 implem.go:104] The path for preference file is /home/mzee1000/.odo/preference.yaml
...
 •  Syncing files into the container  ...
I0310 12:23:44.268307   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile, destBase: ., destFile: .
I0310 12:23:44.268362   17253 file_indexer.go:437] Corrected destinations: base: . file: .
I0310 12:23:44.271245   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/.git, destBase: ., destFile: .git
I0310 12:23:44.271286   17253 file_indexer.go:437] Corrected destinations: base: . file: .git
I0310 12:23:44.272655   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/.gitignore, destBase: ., destFile: .gitignore
I0310 12:23:44.272687   17253 file_indexer.go:437] Corrected destinations: base: . file: .gitignore
I0310 12:23:44.274091   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/.gitignore
I0310 12:23:44.274141   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/.odo, destBase: ., destFile: .odo
I0310 12:23:44.274155   17253 file_indexer.go:437] Corrected destinations: base: . file: .odo
I0310 12:23:44.275447   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/.odo
I0310 12:23:44.275502   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/devfile.yaml, destBase: ., destFile: devfile.yaml
I0310 12:23:44.275521   17253 file_indexer.go:437] Corrected destinations: base: . file: devfile.yaml
I0310 12:23:44.276807   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/devfile.yaml
I0310 12:23:44.276845   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear, destBase: ., destFile: ear
I0310 12:23:44.276858   17253 file_indexer.go:437] Corrected destinations: base: . file: ear
I0310 12:23:44.278056   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear
I0310 12:23:44.278184   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/pom.xml, destBase: ., destFile: ear/pom.xml
I0310 12:23:44.278205   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/pom.xml
I0310 12:23:44.279440   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/pom.xml
I0310 12:23:44.279492   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src, destBase: ., destFile: ear/src
I0310 12:23:44.279507   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src
I0310 12:23:44.280926   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src
I0310 12:23:44.281037   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/main, destBase: ., destFile: ear/src/main
I0310 12:23:44.281060   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/main
I0310 12:23:44.282506   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/main
I0310 12:23:44.282614   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/main/liberty, destBase: ., destFile: ear/src/main/liberty
I0310 12:23:44.282635   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/main/liberty
I0310 12:23:44.284217   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/main/liberty
I0310 12:23:44.284326   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/main/liberty/config, destBase: ., destFile: ear/src/main/liberty/config
I0310 12:23:44.284364   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/main/liberty/config
I0310 12:23:44.285994   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/main/liberty/config
I0310 12:23:44.286102   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/main/liberty/config/server.xml, destBase: ., destFile: ear/src/main/liberty/config/server.xml
I0310 12:23:44.286130   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/main/liberty/config/server.xml
I0310 12:23:44.287750   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/main/liberty/config/server.xml
I0310 12:23:44.287812   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test, destBase: ., destFile: ear/src/test
I0310 12:23:44.287840   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test
I0310 12:23:44.289398   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test
I0310 12:23:44.289560   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java, destBase: ., destFile: ear/src/test/java
I0310 12:23:44.289592   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java
I0310 12:23:44.291346   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java
I0310 12:23:44.291501   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java/it, destBase: ., destFile: ear/src/test/java/it
I0310 12:23:44.291536   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java/it
I0310 12:23:44.293169   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java/it
I0310 12:23:44.293316   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java/it/io, destBase: ., destFile: ear/src/test/java/it/io
I0310 12:23:44.293349   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java/it/io
I0310 12:23:44.296025   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java/it/io
I0310 12:23:44.296199   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java/it/io/openliberty, destBase: ., destFile: ear/src/test/java/it/io/openliberty
I0310 12:23:44.296236   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java/it/io/openliberty
I0310 12:23:44.299166   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java/it/io/openliberty
I0310 12:23:44.299377   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java/it/io/openliberty/guides, destBase: ., destFile: ear/src/test/java/it/io/openliberty/guides
I0310 12:23:44.299411   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java/it/io/openliberty/guides
I0310 12:23:44.302167   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java/it/io/openliberty/guides
I0310 12:23:44.302308   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java/it/io/openliberty/guides/multimodules, destBase: ., destFile: ear/src/test/java/it/io/openliberty/guides/multimodules
I0310 12:23:44.302346   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java/it/io/openliberty/guides/multimodules
I0310 12:23:44.304595   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java/it/io/openliberty/guides/multimodules
I0310 12:23:44.304758   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/src/test/java/it/io/openliberty/guides/multimodules/IT.java, destBase: ., destFile: ear/src/test/java/it/io/openliberty/guides/multimodules/IT.java
I0310 12:23:44.304799   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/src/test/java/it/io/openliberty/guides/multimodules/IT.java
I0310 12:23:44.306827   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/ear/src/test/java/it/io/openliberty/guides/multimodules/IT.java
I0310 12:23:44.306993   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/ear/target, destBase: ., destFile: ear/target
I0310 12:23:44.307024   17253 file_indexer.go:437] Corrected destinations: base: . file: ear/target
I0310 12:23:44.309561   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar, destBase: ., destFile: jar
I0310 12:23:44.309634   17253 file_indexer.go:437] Corrected destinations: base: . file: jar
I0310 12:23:44.313424   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar
I0310 12:23:44.313719   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/pom.xml, destBase: ., destFile: jar/pom.xml
I0310 12:23:44.313761   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/pom.xml
I0310 12:23:44.318018   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/pom.xml
I0310 12:23:44.318069   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src, destBase: ., destFile: jar/src
I0310 12:23:44.318083   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src
I0310 12:23:44.319466   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src
I0310 12:23:44.319645   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main, destBase: ., destFile: jar/src/main
I0310 12:23:44.319670   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main
I0310 12:23:44.320979   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main
I0310 12:23:44.321081   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java, destBase: ., destFile: jar/src/main/java
I0310 12:23:44.321102   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java
I0310 12:23:44.322656   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java
I0310 12:23:44.322760   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java/io, destBase: ., destFile: jar/src/main/java/io
I0310 12:23:44.322784   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java/io
I0310 12:23:44.324344   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java/io
I0310 12:23:44.324546   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java/io/openliberty, destBase: ., destFile: jar/src/main/java/io/openliberty
I0310 12:23:44.324596   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java/io/openliberty
I0310 12:23:44.326148   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java/io/openliberty
I0310 12:23:44.326283   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java/io/openliberty/guides, destBase: ., destFile: jar/src/main/java/io/openliberty/guides
I0310 12:23:44.326312   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java/io/openliberty/guides
I0310 12:23:44.327741   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java/io/openliberty/guides
I0310 12:23:44.327863   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java/io/openliberty/guides/multimodules, destBase: ., destFile: jar/src/main/java/io/openliberty/guides/multimodules
I0310 12:23:44.327891   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java/io/openliberty/guides/multimodules
I0310 12:23:44.329260   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java/io/openliberty/guides/multimodules
I0310 12:23:44.329378   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java/io/openliberty/guides/multimodules/lib, destBase: ., destFile: jar/src/main/java/io/openliberty/guides/multimodules/lib
I0310 12:23:44.329406   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java/io/openliberty/guides/multimodules/lib
I0310 12:23:44.330768   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java/io/openliberty/guides/multimodules/lib
I0310 12:23:44.330903   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/src/main/java/io/openliberty/guides/multimodules/lib/Converter.java, destBase: ., destFile: jar/src/main/java/io/openliberty/guides/multimodules/lib/Converter.java
I0310 12:23:44.330933   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/src/main/java/io/openliberty/guides/multimodules/lib/Converter.java
I0310 12:23:44.332422   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/jar/src/main/java/io/openliberty/guides/multimodules/lib/Converter.java
I0310 12:23:44.332577   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/jar/target, destBase: ., destFile: jar/target
I0310 12:23:44.332616   17253 file_indexer.go:437] Corrected destinations: base: . file: jar/target
I0310 12:23:44.334057   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/pom.xml, destBase: ., destFile: pom.xml
I0310 12:23:44.334089   17253 file_indexer.go:437] Corrected destinations: base: . file: pom.xml
I0310 12:23:44.335574   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/pom.xml
I0310 12:23:44.335622   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war, destBase: ., destFile: war
I0310 12:23:44.335651   17253 file_indexer.go:437] Corrected destinations: base: . file: war
I0310 12:23:44.337116   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war
I0310 12:23:44.337281   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/pom.xml, destBase: ., destFile: war/pom.xml
I0310 12:23:44.337307   17253 file_indexer.go:437] Corrected destinations: base: . file: war/pom.xml
I0310 12:23:44.338738   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/pom.xml
I0310 12:23:44.338796   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src, destBase: ., destFile: war/src
I0310 12:23:44.338824   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src
I0310 12:23:44.340591   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src
I0310 12:23:44.340793   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main, destBase: ., destFile: war/src/main
I0310 12:23:44.340833   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main
I0310 12:23:44.343696   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main
I0310 12:23:44.343911   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java, destBase: ., destFile: war/src/main/java
I0310 12:23:44.343948   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java
I0310 12:23:44.346341   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java
I0310 12:23:44.346557   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java/io, destBase: ., destFile: war/src/main/java/io
I0310 12:23:44.346598   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java/io
I0310 12:23:44.348957   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java/io
I0310 12:23:44.349099   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java/io/openliberty, destBase: ., destFile: war/src/main/java/io/openliberty
I0310 12:23:44.349124   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java/io/openliberty
I0310 12:23:44.350587   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java/io/openliberty
I0310 12:23:44.350821   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java/io/openliberty/guides, destBase: ., destFile: war/src/main/java/io/openliberty/guides
I0310 12:23:44.350842   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java/io/openliberty/guides
I0310 12:23:44.352303   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java/io/openliberty/guides
I0310 12:23:44.352406   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java/io/openliberty/guides/multimodules, destBase: ., destFile: war/src/main/java/io/openliberty/guides/multimodules
I0310 12:23:44.352428   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java/io/openliberty/guides/multimodules
I0310 12:23:44.353996   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java/io/openliberty/guides/multimodules
I0310 12:23:44.354103   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java/io/openliberty/guides/multimodules/web, destBase: ., destFile: war/src/main/java/io/openliberty/guides/multimodules/web
I0310 12:23:44.354131   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java/io/openliberty/guides/multimodules/web
I0310 12:23:44.355462   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java/io/openliberty/guides/multimodules/web
I0310 12:23:44.355610   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/java/io/openliberty/guides/multimodules/web/HeightsBean.java, destBase: ., destFile: war/src/main/java/io/openliberty/guides/multimodules/web/HeightsBean.java
I0310 12:23:44.355635   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/java/io/openliberty/guides/multimodules/web/HeightsBean.java
I0310 12:23:44.357427   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/java/io/openliberty/guides/multimodules/web/HeightsBean.java
I0310 12:23:44.357571   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/webapp, destBase: ., destFile: war/src/main/webapp
I0310 12:23:44.357605   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/webapp
I0310 12:23:44.359720   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/webapp
I0310 12:23:44.359885   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/webapp/WEB-INF, destBase: ., destFile: war/src/main/webapp/WEB-INF
I0310 12:23:44.359913   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/webapp/WEB-INF
I0310 12:23:44.361577   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/webapp/WEB-INF
I0310 12:23:44.361701   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/webapp/WEB-INF/web.xml, destBase: ., destFile: war/src/main/webapp/WEB-INF/web.xml
I0310 12:23:44.361729   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/webapp/WEB-INF/web.xml
I0310 12:23:44.363587   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/webapp/WEB-INF/web.xml
I0310 12:23:44.363683   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/webapp/heights.jsp, destBase: ., destFile: war/src/main/webapp/heights.jsp
I0310 12:23:44.363724   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/webapp/heights.jsp
I0310 12:23:44.365519   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/webapp/heights.jsp
I0310 12:23:44.365561   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/src/main/webapp/index.jsp, destBase: ., destFile: war/src/main/webapp/index.jsp
I0310 12:23:44.365577   17253 file_indexer.go:437] Corrected destinations: base: . file: war/src/main/webapp/index.jsp
I0310 12:23:44.371763   17253 file_indexer.go:487] file added: /home/mzee1000/workbench/work/odo/testing/multi-module-devfile/war/src/main/webapp/index.jsp
I0310 12:23:44.372311   17253 file_indexer.go:431] recursiveTar arguments: srcBase: /home/mzee1000/workbench/work/odo/testing, srcFile: multi-module-devfile/war/target, destBase: ., destFile: war/target
I0310 12:23:44.372376   17253 file_indexer.go:437] Corrected destinations: base: . file: war/target
...
```

You should be able to see that there is no file inside target directorues are being synced. The only appearance of target in the sync logs is for the check of the target paths like `war/target` above for example.

mohammedzee1000 commented:
Ok finally got ci to go green the other git-ignore we tried works but there is 1 case that is not covered as stated in`https://github.com/sabhiram/go-gitignore/blob/master/ignore.go#L92` `TODO: Handle [Rule 4] which negates the match for patterns leading with "!"`

All that remains now is to add tests for actual case that caused this problem

mohammedzee1000 commented:
prow fails are related to yum

mohammedzee1000 commented:
/retest

feloy commented on pkg/util/util.go:
```diff
@@ -99,6 +106,27 @@ func NamespaceKubernetesObject(componentName string, applicationName string) (st
+		// the relative paths in the glob expressions need to be converted to absolute paths
+		var rel string
+		var err error
+		globExp = strings.ReplaceAll(globExp, "*", "_starchar_")
```
If we need to use such a hack, it is because we didn't understand how to use the library, or it is buggy, or ...

Could you please create a new PR with the other gitignore library, so we can see which one is the best in our case?

mohammedzee1000 replied:
The problem is filepath.Rel does not calculate relative path when there are symbols like *

valaparthvi commented:
Tested this with .gitignore and .odoignore, and the functionality looks good to me.
/approve

valaparthvi commented:
@mohammedzee1000 I noticed that you are using the multi-module project as it from Scott's repo. I think you could instead create a temporary folder/s during the test setup, add it to .gitignore/.odoignore and see if it shows up on odo exec, I do not think you need to clone the whole project locally.

mohammedzee1000 commented:
The errors seem unrelated to code changes. whats going on here @anandrkskd ? Also whats with html logs for windows it looks like s**t

mohammedzee1000 commented:
Based on comm with @anandrkskd the ci going green is blocked on https://github.com/redhat-developer/odo/pull/5530

mohammedzee1000 commented:
cluster provision failure
/test unit

mohammedzee1000 commented:
> @mohammedzee1000 I noticed that you are using the multi-module project as it is from Scott's repo. I think you could instead create a temporary folder/s during the test setup, add it to .gitignore/.odoignore and see if it shows up on odo exec, I do not think you need to clone the whole project locally.

Got it I figured out how the testing into UTs so I have removed int test that i had added completely

mohammedzee1000 commented:
The windows tests are still failing even after rebase @anandrkskd  seems to be preference related :/

anandrkskd commented:
> The windows tests are still failing even after rebase @anandrkskd seems to be preference related :/

I will try to fix it ASAP.

anandrkskd commented:
Hey @mohammedzee1000 I think this is a problem [here](https://github.com/redhat-developer/odo/pull/5541/files#diff-be3784f44e255e6d6d44c32b41432205014600759319e14a072a0bc10ed1b3c9R138), here we are retunring an err which is not constant(changes on basis of underlying OS used), this is the reason why it failed on your PR. I don't think this is a error is due to CI. 

@valaparthvi can we return a constant error from our side [here](https://github.com/redhat-developer/odo/pull/5541/files#diff-be3784f44e255e6d6d44c32b41432205014600759319e14a072a0bc10ed1b3c9R138) and use that for testing?

mohammedzee1000 commented:
> Hey @mohammedzee1000 I think this is a problem [here](https://github.com/redhat-developer/odo/pull/5541/files#diff-be3784f44e255e6d6d44c32b41432205014600759319e14a072a0bc10ed1b3c9R138), here we are retunring an err which is not constant(changes on basis of underlying OS used), this is the reason why it failed on your PR. I don't think this is a error is due to CI.
> 
> @valaparthvi can we return a constant error from our side [here](https://github.com/redhat-developer/odo/pull/5541/files#diff-be3784f44e255e6d6d44c32b41432205014600759319e14a072a0bc10ed1b3c9R138) and use that for testing?

Maybe an `fmt.Errof("constant msg: %w", err)`. That way, we can check for constant string,
Although why are we checking for an exact error string if multiple different errors can occor

feloy commented on pkg/util/file_indexer.go:
```diff
@@ -365,10 +366,15 @@ func runIndexerWithExistingFileIndex(directory string, ignoreRules []string, rem
+				relx, err := filepath.Rel(directory, fileAbsolutePath)
 				if err != nil {
-					return IndexerRet{}, err
+					continue // TODO
```
I think you can revert to `return IndexRet{}, err`

feloy commented on pkg/util/file_indexer.go:
```diff
@@ -444,23 +450,25 @@ func recursiveChecker(pathOptions recursiveCheckerPathOptions, ignoreRules []str
+		rel, err := filepath.Rel(pathOptions.directory, matchedPath)
 		if err != nil {
-			return IndexerRet{}, err
+			continue // TODO
```
same here

feloy commented on pkg/util/util.go:
```diff
@@ -32,6 +32,13 @@ import (
 	"k8s.io/klog"
 )
 
+const (
```
I think this block comes from a rebase. It seems you can remove it

mohammedzee1000 replied:
Ok let me see if it breaks anything by removing it

feloy commented:
LGTM. Thanks for this work @mohammedzee1000
